### DISCUSSION TITLE
Can I set environment variables dynamically for each preview deployment?

### DISCUSSION THREAD

**Discussion by jasongoetz** [vercel/vercel#8543] [Help] [answered]
I'm gotten stuck when it comes to building a vercel preview deployment through GitHub Actions. Each preview deployment is going to have a separate GraphQL URL that it talks to. So I need each preview deployment to have an environment variable (e.g. NEXT_PUBLIC_GRAPHQL_SERVER) that points to its own dedicated GraphQL server. 

It seems to me that environment variables can only be defined as static values on preview environments. How would I be able to set up my vercel deployments so they get this value dynamically?

UPDATE:

This document explains how to build and deploy via GHA: https://vercel.com/guides/how-can-i-use-github-actions-with-vercel

When I use that workflow, the system provided variables are never set though. For example, process.env.NEXT_PUBLIC_VERCEL_URL is undefined after running `vercel build` and `vercel deploy`. How do I get my system provided variables?

**Comment by amyegan** [comment, marked as answer]
Thanks for reaching out @jasongoetz! I believe the [Vercel for Git](https://vercel.com/docs/concepts/git) integrations will automatically set the framework environment variables during build, but I'm not sure that happens with prebuilt deployments. Couple of ideas to try:
- Can you get values from non-prefixed [System Environment Variables](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables), such as `VERCEL_ENV`?
- Try adding a `NEXT_PUBLIC_` prefixed environment variable from your Action using [Vercel CLI env command](https://vercel.com/docs/cli/env#)

If it still isn't working for you, can you share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example)? That way I can get a better look at what's going wrong for you.

**Reply by jasongoetz** [reply]
Thanks @amyegan - your ideas helped. The `NEXT_PUBLIC_` variables are not being added with prebuilt deployments like you suspected, but I was able to add them directly via the CLI env command.
